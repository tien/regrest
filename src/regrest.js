/**
 * Config options
 * @typedef {Object.<string, *>} Config
 * @property {string} [method = "GET"] - HTTP request method
 * @property {string} [url] - The url
 * @property {Object.<string, string>} [headers = {}] - The request headers
 * @property {Object.<string, *>} [params] - The request query
 * @property {*} [data = null] - The data to be sent
 * @property {number} [maxRedirects = 5] - Maximum redirects before error is thrown
 * @property {boolean} [withCredentials = false] - Whether cross-site Access-Control requests should be made using credentials
 *
 * Regrest response object
 * @typedef {Object.<string, *>} Response
 * @property {number} status - The response status code
 * @property {string} statusText - The response status text
 * @property {Object.<string, string>} headers - The response headers
 * @property {string} text - The response raw text
 * @property {Object} json - The response parsed as JSON
 * @property {Blob|any} arrayBuffer - The response as an Blob on browser or Buffer on Node js
 * @property {Blob} blob - The response as a Blob object
 */

const ENVIRONMENTS = Object.freeze({ BROWSER: 0, NODE: 1, UNKNOWN: 2 });

// Detect whether instance is ran in browser or on node js
const ENV =
  typeof window === "object" && typeof window.document === "object"
    ? ENVIRONMENTS.BROWSER
    : typeof module === "object" && module && typeof module.exports === "object"
    ? ENVIRONMENTS.NODE
    : ENVIRONMENTS.UNKNOWN;

class NetworkError extends Error {
  constructor(message, request, response) {
    super(message);
    this.name = this.constructor.name;
    this.response = response;
    this.request = request;
    if (typeof Error.captureStackTrace === "function") {
      Error.captureStackTrace(this, this.constructor);
    } else {
      this.stack = new Error(message).stack;
    }
  }
}

function Regrest() {
  switch (ENV) {
    case ENVIRONMENTS.BROWSER:
      this.requestAdapter = xhrAdapter.bind(this);
      break;
    case ENVIRONMENTS.NODE:
      this.nodeAdapters = {
        http: require("follow-redirects").http,
        https: require("follow-redirects").https,
      };
      this.requestAdapter = httpAdapter.bind(this);
      break;
    default:
      throw new Error("Unsupported environment");
  }
}

/**
 * @param {Config} config
 * @param {string} config.url - The url
 * @returns {Promise<Response>}
 * @memberof Regrest
 */
Regrest.prototype.request = function ({
  method = "GET",
  url,
  headers = {},
  params,
  data = null,
  maxRedirects = 5,
  withCredentials = false,
}) {
  // Generate query string and join it with url
  url = `${url}${
    params
      ? `?${Object.entries(params)
          .map(([key, value]) => `${key}=${value}`)
          .join("&")}`
      : ""
  }`;
  return this.requestAdapter(
    method,
    url,
    data,
    headers,
    maxRedirects,
    withCredentials
  );
};

// Convenience methods
/**
 * @param {string} url - The url
 * @param {Config} [config] - Config
 * @returns {Promise<Response>}
 * @memberof Regrest
 */
["get", "head", "delete", "options"].forEach(
  (method) =>
    (Regrest.prototype[method] = function (url, config) {
      return this.request({ ...config, method: method.toUpperCase(), url });
    })
);

/**
 * @param {string} url - The url
 * @param {*} [data] - The data to be sent
 * @param {Config} [config] - Config
 * @returns {Promise<Response>}
 * @memberof Regrest
 */
["post", "put", "patch"].forEach(
  (method) =>
    (Regrest.prototype[method] = function (url, data, config) {
      return this.request({
        ...config,
        method: method.toUpperCase(),
        url,
        data,
      });
    })
);

// Export
if (
  typeof module === "object" &&
  module &&
  typeof module.exports === "object"
) {
  /**
   * Expose Regrest as module.exports in loaders
   * that implement the Node module pattern (including browserify)
   */
  module.exports = new Regrest();
} else {
  // Otherwise expose Regrest to the global object
  window.regrest = new Regrest();
}

// Unexposed helper methods and adapters
function xhrAdapter(requestType, url, body, headers, _, withCredentials) {
  return new Promise((resolve, reject) => {
    const request = new XMLHttpRequest();
    request.open(requestType, url, true);
    Object.entries(headers).forEach(([key, value]) =>
      request.setRequestHeader(key, value)
    );
    withCredentials && (request.withCredentials = true);
    request.onload = function () {
      const contentType = (this.getResponseHeader("Content-Type") || "")
        .split(";")[0]
        .trim();
      const response = {
        status: this.status,
        statusText: this.statusText,
        headers: {
          ...this.getAllResponseHeaders()
            .trim()
            .split(/[\r\n]+/)
            .map((header) => header.split(": "))
            .reduce(
              (obj, [key, value]) => ({ ...obj, [key.toLowerCase()]: value }),
              {}
            ),
        },
        text: this.responseText,
        get json() {
          delete this.json;
          return (this.json = JSON.parse(this.text));
        },
        get blob() {
          delete this.blob;
          return (this.blob = new Blob([request.response], {
            type: contentType,
          }));
        },
        get arrayBuffer() {
          delete this.arrayBuffer;
          return (this.arrayBuffer = this.blob);
        },
      };
      if (this.status >= 200 && this.status < 400) {
        resolve(response);
      } else {
        reject(
          new NetworkError(
            `${this.status} ${this.statusText}`,
            request,
            response
          )
        );
      }
    };
    request.onerror = function () {
      reject(new NetworkError("connection error", request));
    };
    request.send(body);
  });
}

function httpAdapter(requestType, url, body, headers, maxRedirects) {
  return new Promise((resolve, reject) => {
    const parsedUrl = new URL(url);
    const options = {
      host: parsedUrl.host,
      path: `${parsedUrl.pathname}${parsedUrl.search}`,
      method: requestType,
      headers,
      maxRedirects,
    };
    const req = this.nodeAdapters[parsedUrl.protocol.slice(0, -1)].request(
      options,
      (res) => {
        const response = {
          status: res.statusCode,
          statusText: res.statusMessage,
          headers: res.headers,
          arrayBuffer: [],
          get text() {
            delete this.text;
            return (this.text = this.arrayBuffer.toString("utf-8"));
          },
          get json() {
            delete this.json;
            return (this.json = JSON.parse(this.text));
          },
          get blob() {
            if (typeof Blob !== "function") {
              throw new Error("Please include Blob polyfill for Node.js");
            }
            const contentType = this.headers["content-type"] || "";
            delete this.blob;
            return (this.blob = new Blob([new Uint8Array(this.arrayBuffer)], {
              type: contentType.split(";")[0].trim(),
            }));
          },
        };
        res.on("data", (chunk) => response.arrayBuffer.push(chunk));
        res.on("end", () => {
          response.arrayBuffer = Buffer.concat(response.arrayBuffer);
          if (res.statusCode >= 200 && res.statusCode < 400) {
            resolve(response);
          } else {
            reject(
              new NetworkError(
                `${res.statusCode} ${res.statusMessage}`,
                req,
                response
              )
            );
          }
        });
      }
    );
    req.on("error", (e) => reject(new NetworkError(e.message, req)));
    body && req.write(body);
    req.end();
  });
}
