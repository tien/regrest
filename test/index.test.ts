import regrest, { Response } from "../src";

const expectResponseSnapshot = (received: Response, statusCode: number) => {
  expect(received.status).toBe(statusCode);
  expect(received.text).toMatchSnapshot();
  expect(received.json).toMatchSnapshot();
};

describe("regrest", () => {
  test("GET", async () => {
    const response = await regrest.get(
      "https://jsonplaceholder.typicode.com/posts/1"
    );

    expectResponseSnapshot(response, 200);
  });

  test("GET with params", async () => {
    const response = await regrest.get(
      "http://jsonplaceholder.typicode.com/comments",
      {
        params: { postId: 1 },
      }
    );

    expectResponseSnapshot(response, 200);
  });

  test("POST", async () => {
    const response = await regrest.post(
      "https://jsonplaceholder.typicode.com/posts",
      JSON.stringify({
        title: "foo",
        body: "bar",
        userId: 1,
      }),
      {
        headers: {
          "Content-type": "application/json; charset=UTF-8",
        },
      }
    );

    expectResponseSnapshot(response, 201);
  });

  test("PUT", async () => {
    const response = await regrest.put(
      "https://jsonplaceholder.typicode.com/posts/1",
      JSON.stringify({
        id: 1,
        title: "foo",
        body: "bar",
        userId: 1,
      }),
      {
        headers: {
          "Content-type": "application/json; charset=UTF-8",
        },
        maxRedirects: 10,
      }
    );

    expectResponseSnapshot(response, 200);
  });

  test("PATCH", async () => {
    const response = await regrest.patch(
      "https://jsonplaceholder.typicode.com/posts/1",
      JSON.stringify({
        title: "foo",
      }),
      {
        headers: {
          "Content-type": "application/json; charset=UTF-8",
        },
      }
    );

    expectResponseSnapshot(response, 200);
  });

  test("DELETE", async () => {
    const response = await regrest.delete(
      "https://jsonplaceholder.typicode.com/posts/1"
    );

    expectResponseSnapshot(response, 200);
  });

  test("errors handling", async () => {
    try {
      await regrest.post("https://jsonplaceholder.typicode.com/rubbish");
      expect(true).toBe(false);
    } catch (error) {
      expect(error).toMatchInlineSnapshot(`[NetworkError: 404 Not Found]`);
    }
  });
});
