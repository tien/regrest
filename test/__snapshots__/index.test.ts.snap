// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`regrest DELETE 1`] = `"{}"`;

exports[`regrest DELETE 2`] = `Object {}`;

exports[`regrest GET 1`] = `
"{
  \\"userId\\": 1,
  \\"id\\": 1,
  \\"title\\": \\"sunt aut facere repellat provident occaecati excepturi optio reprehenderit\\",
  \\"body\\": \\"quia et suscipit\\\\nsuscipit recusandae consequuntur expedita et cum\\\\nreprehenderit molestiae ut ut quas totam\\\\nnostrum rerum est autem sunt rem eveniet architecto\\"
}"
`;

exports[`regrest GET 2`] = `
Object {
  "body": "quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto",
  "id": 1,
  "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
  "userId": 1,
}
`;

exports[`regrest GET with params 1`] = `
"[
  {
    \\"postId\\": 1,
    \\"id\\": 1,
    \\"name\\": \\"id labore ex et quam laborum\\",
    \\"email\\": \\"Eliseo@gardner.biz\\",
    \\"body\\": \\"laudantium enim quasi est quidem magnam voluptate ipsam eos\\\\ntempora quo necessitatibus\\\\ndolor quam autem quasi\\\\nreiciendis et nam sapiente accusantium\\"
  },
  {
    \\"postId\\": 1,
    \\"id\\": 2,
    \\"name\\": \\"quo vero reiciendis velit similique earum\\",
    \\"email\\": \\"Jayne_Kuhic@sydney.com\\",
    \\"body\\": \\"est natus enim nihil est dolore omnis voluptatem numquam\\\\net omnis occaecati quod ullam at\\\\nvoluptatem error expedita pariatur\\\\nnihil sint nostrum voluptatem reiciendis et\\"
  },
  {
    \\"postId\\": 1,
    \\"id\\": 3,
    \\"name\\": \\"odio adipisci rerum aut animi\\",
    \\"email\\": \\"Nikita@garfield.biz\\",
    \\"body\\": \\"quia molestiae reprehenderit quasi aspernatur\\\\naut expedita occaecati aliquam eveniet laudantium\\\\nomnis quibusdam delectus saepe quia accusamus maiores nam est\\\\ncum et ducimus et vero voluptates excepturi deleniti ratione\\"
  },
  {
    \\"postId\\": 1,
    \\"id\\": 4,
    \\"name\\": \\"alias odio sit\\",
    \\"email\\": \\"Lew@alysha.tv\\",
    \\"body\\": \\"non et atque\\\\noccaecati deserunt quas accusantium unde odit nobis qui voluptatem\\\\nquia voluptas consequuntur itaque dolor\\\\net qui rerum deleniti ut occaecati\\"
  },
  {
    \\"postId\\": 1,
    \\"id\\": 5,
    \\"name\\": \\"vero eaque aliquid doloribus et culpa\\",
    \\"email\\": \\"Hayden@althea.biz\\",
    \\"body\\": \\"harum non quasi et ratione\\\\ntempore iure ex voluptates in ratione\\\\nharum architecto fugit inventore cupiditate\\\\nvoluptates magni quo et\\"
  }
]"
`;

exports[`regrest GET with params 2`] = `
Array [
  Object {
    "body": "laudantium enim quasi est quidem magnam voluptate ipsam eos
tempora quo necessitatibus
dolor quam autem quasi
reiciendis et nam sapiente accusantium",
    "email": "Eliseo@gardner.biz",
    "id": 1,
    "name": "id labore ex et quam laborum",
    "postId": 1,
  },
  Object {
    "body": "est natus enim nihil est dolore omnis voluptatem numquam
et omnis occaecati quod ullam at
voluptatem error expedita pariatur
nihil sint nostrum voluptatem reiciendis et",
    "email": "Jayne_Kuhic@sydney.com",
    "id": 2,
    "name": "quo vero reiciendis velit similique earum",
    "postId": 1,
  },
  Object {
    "body": "quia molestiae reprehenderit quasi aspernatur
aut expedita occaecati aliquam eveniet laudantium
omnis quibusdam delectus saepe quia accusamus maiores nam est
cum et ducimus et vero voluptates excepturi deleniti ratione",
    "email": "Nikita@garfield.biz",
    "id": 3,
    "name": "odio adipisci rerum aut animi",
    "postId": 1,
  },
  Object {
    "body": "non et atque
occaecati deserunt quas accusantium unde odit nobis qui voluptatem
quia voluptas consequuntur itaque dolor
et qui rerum deleniti ut occaecati",
    "email": "Lew@alysha.tv",
    "id": 4,
    "name": "alias odio sit",
    "postId": 1,
  },
  Object {
    "body": "harum non quasi et ratione
tempore iure ex voluptates in ratione
harum architecto fugit inventore cupiditate
voluptates magni quo et",
    "email": "Hayden@althea.biz",
    "id": 5,
    "name": "vero eaque aliquid doloribus et culpa",
    "postId": 1,
  },
]
`;

exports[`regrest PATCH 1`] = `
"{
  \\"userId\\": 1,
  \\"id\\": 1,
  \\"title\\": \\"foo\\",
  \\"body\\": \\"quia et suscipit\\\\nsuscipit recusandae consequuntur expedita et cum\\\\nreprehenderit molestiae ut ut quas totam\\\\nnostrum rerum est autem sunt rem eveniet architecto\\"
}"
`;

exports[`regrest PATCH 2`] = `
Object {
  "body": "quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto",
  "id": 1,
  "title": "foo",
  "userId": 1,
}
`;

exports[`regrest POST 1`] = `
"{
  \\"title\\": \\"foo\\",
  \\"body\\": \\"bar\\",
  \\"userId\\": 1,
  \\"id\\": 101
}"
`;

exports[`regrest POST 2`] = `
Object {
  "body": "bar",
  "id": 101,
  "title": "foo",
  "userId": 1,
}
`;

exports[`regrest PUT 1`] = `
"{
  \\"id\\": 1,
  \\"title\\": \\"foo\\",
  \\"body\\": \\"bar\\",
  \\"userId\\": 1
}"
`;

exports[`regrest PUT 2`] = `
Object {
  "body": "bar",
  "id": 1,
  "title": "foo",
  "userId": 1,
}
`;
